### ISSUE TITLE
Importing weights from a model trained in Tensorflow to CMSIS NN 

### ISSUE THREAD

**Issue by petitcoeur9000** [ARM-software/CMSIS_5#343]
Hi,

I am interested in training a CNN in Tensorflow and then importing the trained weights to a CNN implemented with the CMSIS NN library. Is it doable? Is there any need to preprocess the weights from the Tensorflow model before importing it?

Thanks

**Comment by liangzhen-lai** [comment]
Hi @petitcoeur9000 ,

Yes, it is doable. We have a few examples of model trained in tensorflow and ported to the CMSIS-NN library. It will require quantization on the weights and activations. You can refer to this for more details:
https://github.com/ARM-software/ML-KWS-for-MCU/blob/master/Deployment/Quant_guide.md

We are also working on some simpler (e.g., semi-automatic) ways for people to port Tensorflow model with CMSIS-NN. For now, you can refer to the documents mentioned above. Let me know if this answers your question.

Thanks,
Liangzhen

**Comment by petitcoeur9000** [comment]
Thanks for the answer! the tensorflow porting examples you are referring to are all contained in the ML-KWS-for-MCU repo or are they also somewhere else?

**Comment by liangzhen-lai** [comment]
Also, if you have fully-connected layers in your network and want to use the kernel with weight reordering (i.e., arm_fully_connected_**_opt), you can refer to this script for weight reordering:
[Script link](https://github.com/ARM-software/CMSIS_5/blob/develop/CMSIS/NN/Scripts/NNFunctions/fully_connected_opt_weight_generation.py)

**Comment by petitcoeur9000** [comment]
thanks!

**Comment by lishunde** [comment]
It's great to see this issues,I also want to know how to generate the data in  [https://github.com/ARM-software/CMSIS_5/blob/develop/CMSIS/NN/Examples/ARM/arm_nn_examples/cifar10/arm_nnexamples_cifar10_shift.h](url)

**Comment by petitcoeur9000** [comment]
Hi @lishunde ,

see #327 to understand how to generate the shift header.

**Comment by fengzi0** [comment]
Hi @liangzhen-lai ， i'm trying to CNN model from TensorFlow to CMSIS_NN，but the order of CNN's weight is puzzling me，is it N(output_channel)H(eight)W(idth)C(input_channel)? or HWCN as TensorFlow? Also about full connection layer, is it N(output_channel)C(input_channel) or CN?

**Comment by liangzhen-lai** [comment]
Hi @fengzi0 , it should be NHWC and NC.

**Comment by sbochkar** [comment]
Hi @liangzhen-lai ,

Just want to clarify the CNN's weight format that arm_convolve_HWC_q7 expects.

Does it expect a flattened array of weights where the first index varies the slowest and the last index varies the fastest?

Example:
Suppose I have 29 5x5 conv. filters acting on 3 input channels.
In Caffe this would be an array of shape (29, 3, 5, 5). NCHW
To make this work with CMSIS-NN, I should first rearrange this matrix into (29, 5, 5, 3) [NHWC]. And then flatten using row major transform. e.g. numpy.ravel(). (e.g. last index varies fastest)

Can you confirm this is correct? Thank you very much for your time.

**Comment by liangzhen-lai** [comment]
@sbochkar , Yes, the weight ordering is correct. In the cnn example in CMSIS-NN, we use `numpy.reshape([N*H*W*C])` to flatten the weights. I guess `numpy.ravel()` should behave the same.

**Comment by iabdalkader** [comment]
@liangzhen-lai  Hi, I'm also trying to import Caffe models. Can you share the Caffe model you used in the example so I have something to compare to ?

**Comment by iabdalkader** [comment]
>> We are also working on some simpler (e.g., semi-automatic) ways for people to port Tensorflow model with CMSIS-NN.

@liangzhen-lai  Actually never mind the model, have you made any progress on this ?

**Comment by yyz1985** [comment]
> Also, if you have fully-connected layers in your network and want to use the kernel with weight reordering (i.e., arm_fully_connected_**_opt), you can refer to this script for weight reordering:
> [Script link](https://github.com/ARM-software/CMSIS_5/blob/develop/CMSIS/NN/Scripts/NNFunctions/fully_connected_opt_weight_generation.py)

what if n is not multiply of 4, like 10?

**Comment by csukuangfj** [comment]
@yyz1985

> what if n is not multiply of 4, like 10?

It works as you expected. Please read the code for more details.
